server/src/main/java/org/opensearch/index/MergePolicyConfig.java

```diff
@@ -122,11 +129,37 @@
  * possibly either increase the <code>max_merged_segment</code> or issue an optimize
  * call for the index (try and aim to issue it on a low traffic time).
  *
+ /**
+ * <p>
+ * The LogByteSizeMergePolicy is an alternative merge policy designed to optimize the merging of segments in scenarios
+ * with time-based data.
+ * While the TieredMergePolicy is the default choice, the LogByteSizeMergePolicy can be configured
+ * as the default merge policy for time-based data using the <code>index.datastream_merge.policy</code> setting.
+ *
+ * <p>
+ * Unlike the TieredMergePolicy, which prioritizes merging segments of equal sizes, the LogByteSizeMergePolicy
+ * specializes in merging adjacent segments efficiently.
+ * This characteristic makes it particularly well-suited for range queries on time-based data.
+ * Typically, adjacent segments in time-based data often contain documents with similar timestamps.
+ * When these segments are merged, the resulting segment covers a range of timestamps with reduced overlap compared
+ * to the adjacent segments. This reduced overlap remains even as segments grow older and larger,
+ * which can significantly benefit range queries on timestamps.
+ *
+ * <p>
+ * In contrast, the TieredMergePolicy does not honor this timestamp range optimization. It focuses on merging segments
+ * of equal sizes and does not consider adjacency. Consequently, as segments grow older and larger,
+ * the overlap of timestamp ranges among adjacent segments managed by TieredMergePolicy can increase.
+ * This can lead to inefficiencies in range queries on timestamps, as the number of segments to be scanned
+ * within a given timestamp range could become high.
+ *
  * @opensearch.internal
  */
 
 public final class MergePolicyConfig {
```

Comment:
It seems like the data streams benefit from `LogByteSizeMergePolicy` but it could be applied to any other indices, right? 

The implementation bundles all merge policies together (under the same `MergePolicyConfig`) however the index could only have one merge policy at a time (+ flush on merge). I think we should introduce:
 - `index.merge.policy=<policy>` to name the merge policy for the index (default is `tiered`, `log_byte_size` as another option)
 - we need to introduce a generic interface to denote merge policy provider, for example
    ```
    interface MergePolicyProvider {
        MergePolicy getMergePolicy();
    }
    ```
- revert `MergePolicyConfig` and introduce `LogByteSizeMergePolicyConfig`, those two would be separated and both implement `MergePolicyProvider`
- `IndexSettings` would use `MergePolicyProvider` instead of `MergePolicyConfig` directly (it only needs the `getMergePolicy()` anyway

We could keep the default merge policy setting for data stream indices (it should not be an index setting but cluster setting) but with this design, the `IndexSettings` would always maintain the merge policy that is applicable for current index.

Reply:
> It seems like the data streams benefit from LogByteSizeMergePolicy but it could be applied to any other indices, right?

Yes, i should stop calling data streams only as its applicable for any index with `@timestamp` field. I will make the changes appropriately. 

I like your suggestion of having `index.merge.policy`, this would enable us to play with more experimental merge policies tomorrow. And having `MergePolicyProvider` and a cluster level setting for any time-based index goes inline with it. I will incorporate them in next iteration. Thanks for reviewing.

Reply:
Just to save you some time (hopefully), `IndexSettings` has access to `IndexMetadata` that in turn has mappings, we should be able to use that to detect if index is datastream or not.

Reply:
I couldn't find an elegant way to find timestamp based index in IndexSetting. Please check the updated version, it addresses most of the concerns raised here. 

I have introduced 2 settings - `index.merge.policy` (index setting, static) `indices.time_index.merge.policy` (node scoped setting, cluster setting, static). 
`index.merge.policy` will always take the precedence if set. Check `testMergePolicyPrecedence`, let me know if the behavior looks appropriate to you. Thanks.

Reply:
Thanks @rishabhmaurya , understood the complexity, I think overall looks good (to me), a few comments to validation and choosing the right policy, I think we could simplify it

Reply:
Thanks @reta for the review. I would like you to run through the logic of force merges for LogByteSizeMergePolicy. 
For tiered merge policy, since it doesn't honors `setMaxMergedSegmentMB` for force merges (https://issues.apache.org/jira/browse/LUCENE-7976) we have a custom `OpenSearchTieredMergePolicy`. For LogByteSizeMergePolicy it seems to honor this setting thus didn't have to create a wrapper over it.

Reply:
>  For LogByteSizeMergePolicy it seems to honor this setting thus didn't have to create a wrapper over it.

I think this is great, @msfroh do you know about any hidden quirks we should be aware of?

Reply:
So, the issue with TieredMergePolicy is actually that it *does* honor max segment size, even on force merge. The guidance from the Lucene community was "if you want to force merge down to a single segment, increase the max segment size first." Since people had already defined ISM policies to force merge down to a single segment without increasing max segment size, the decision was made to keep the old force merge behavior in Elasticsearch by wrapping the merge policy. (I disagree with that decision when `only_expunge_deletes` was specified, hence https://github.com/opensearch-project/OpenSearch/pull/10036.)

Anyway, `LogByteSizeMergePolicy` gets around this by having two different max segment sizes -- the max segment size for force merge, and the regular max segment size. The max segment size for force merge defaults to `Long.MAX_VALUE`, so @rishabhmaurya is correct that there's no need for the wrapper. 

Interestingly, I was just reading the code for `LogMergePolicy.findForcedDeletesMerges`, and it looks like it *doesn't even look at segment sizes*.  So, if someone is using `LogByteSizeMergePolicy` and they try to force merge deletes, they will almost certainly end up with large segments. On the other hand, I would suggest that anyone who with a use-case that involves deletes or updates to existing docs should not be using `LogByteSizeMergePolicy`.

Reply:
That's a good point. Also, going through the logic for `findForcedDeletesMerges` in  `LogByteSizeMergePolicy`, its going to merge segments more aggressively for delete cases as it doesn't have `pctDeletes` settings to control it.

Reply:
Since we are recommending `LogByteSizeMergePolicy` policy for data streams, which are essentially append only, the deletes should be rather rare.

Reply:
@reta @msfroh I have added outstanding items here before we make LogByteSizeMergePolicy as default for timestamp based indices - https://github.com/opensearch-project/OpenSearch/issues/9241#issuecomment-1735916458
Will resolve this conversation if you both agree.